src/Swarm/TUI/View.hs

```diff
@@ -406,13 +432,14 @@ drawKeyMenu s =
   cheat = s ^. uiState . uiCheatMode
   goal = (s ^. uiState . uiGoal) /= NoGoal
 
-  availRecipes
-    | null (s ^. gameState . availableRecipes) = []
+  notificationKey :: Lens' GameState (Notifications a) -> Text -> Text -> [(KeyHighlight, Text, Text)]
```

Comment:
Not sure why, but without this annotation, ghc fails with `Couldn't match type ‘[a]’ with ‘Int’`. And this needs RankNType to work. Thus I'm not sure it's worth the trouble to use a generic Notifications type, and perhaps it would be better to just duplicate the code. On the other hand, we might want to use this logic for an events log menu too, e.g. when a monster appears, or when a robot gets destroyed.

Reply:
I'm not sure I understand what's going on here in particular, but I do like the generic `Notifications` type.  It does feel like something we might reuse for other purposes later.

Reply:
Anything that takes a lens as an argument is going to require `RankNTypes`, that's not a problem.  The fact that there's a type failure without the annotation makes sense to me (even though I would be hard-pressed to explain the error itself), since GHC will never infer a higher-rank type.

Reply:
I see, thank you so much for the explanations.